hbase-replication/src/main/java/org/apache/hadoop/hbase/replication/ReplicationPeerMigrationTool.java

```diff
@@ -0,0 +1,116 @@
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.replication;
+
+import java.io.IOException;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.conf.Configured;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.hbase.Abortable;
+import org.apache.hadoop.hbase.HBaseConfiguration;
+import org.apache.hadoop.hbase.HBaseInterfaceAudience;
+import org.apache.hadoop.hbase.zookeeper.ZKWatcher;
+import org.apache.hadoop.util.Tool;
+import org.apache.hadoop.util.ToolRunner;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A tool for migrating replication peer data across different replication peer storages.
+ */
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.TOOLS)
+public class ReplicationPeerMigrationTool extends Configured implements Tool {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ReplicationPeerMigrationTool.class);
+
+  public static final String NAME = "migratereppeer";
+
+  public ReplicationPeerMigrationTool(Configuration conf) {
+    super(conf);
+  }
+
+  private ReplicationPeerStorage create(String type, FileSystem fs, ZKWatcher zk) {
+    Configuration conf = getConf();
+    Configuration srcConf = new Configuration(conf);
+    srcConf.set(ReplicationStorageFactory.REPLICATION_PEER_STORAGE_IMPL, type);
+    return ReplicationStorageFactory.getReplicationPeerStorage(fs, zk, srcConf);
+  }
+
+  private ZKWatcher createZKWatcher() throws IOException {
+    return new ZKWatcher(getConf(), getClass().getSimpleName(), new Abortable() {
+
+      private volatile boolean aborted;
+
+      @Override
+      public boolean isAborted() {
+        return aborted;
+      }
+
+      @Override
+      public void abort(String why, Throwable e) {
+        aborted = true;
+        LOG.error(why, e);
+        System.exit(1);
+      }
+    });
+  }
+
+  private void migrate(ReplicationPeerStorage src, ReplicationPeerStorage dst)
+    throws ReplicationException {
+    LOG.info("Start migrating from {} to {}", src.getClass().getSimpleName(),
+      dst.getClass().getSimpleName());
+    for (String peerId : src.listPeerIds()) {
+      LOG.info("Going to migrate {}", peerId);
+      ReplicationPeerConfig peerConfig = src.getPeerConfig(peerId);
+      boolean enabled = src.isPeerEnabled(peerId);
+      SyncReplicationState syncState = src.getPeerSyncReplicationState(peerId);
+      SyncReplicationState newSyncState = src.getPeerNewSyncReplicationState(peerId);
+      if (newSyncState != SyncReplicationState.NONE) {
+        throw new IllegalStateException("Can not migrate peer " + peerId
+          + " as it is in an intermediate state, syncReplicationState is " + syncState
+          + " while newSyncReplicationState is " + newSyncState);
+      }
+      dst.addPeer(peerId, peerConfig, enabled, syncState);
+      LOG.info("Migrated peer {}, peerConfig = '{}', enabled = {}, syncReplicationState = {}",
```

Comment:
That's a good point. I do not think we should delete the peers from src storage, as if we want to do online migration, the peers in the src storage are still in use...

So I think we'd better name it Copy instead of Migration.